etcdserver/snapshot_store.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2015 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package etcdserver
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path"
+
+	"github.com/coreos/etcd/pkg/fileutil"
+	"github.com/coreos/etcd/raft/raftpb"
+	dstorage "github.com/coreos/etcd/storage"
+	"github.com/coreos/etcd/store"
+)
+
+type snapshotV2 struct {
+	// raft snapshot metadata
+	metadata raftpb.SnapshotMetadata
+	// v2 store clone
+	clone store.Store
+}
+
+type snapshotStore struct {
+	// dir to save snapshot data
+	dir string
+	kv  dstorage.KV
+
+	// send empty to reqsnapc to notify the channel receiver to send back latest
+	// snapshot to snapc
+	reqsnapc chan struct{}
+	// a chan to receive the requested snapshot
+	// hub will receive from the chan immediately after it sends empty to reqsnapc
+	snapc chan snapshotV2
+}
+
+func newSnapshotStore(dir string, kv dstorage.KV) *snapshotStore {
+	return &snapshotStore{
+		dir:      dir,
+		kv:       kv,
+		reqsnapc: make(chan struct{}),
+		snapc:    make(chan snapshotV2),
+	}
+}
+
+// Snapshot returns raft snapshot with latest index.
+// TODO: it should return error when there is one in flight.
+// TODO: the function should be unblock and finish in 1ms. It may use long time
+// if v2 store is big.
+func (sh *snapshotStore) Snapshot() raftpb.Snapshot {
+	// ask to generate latest v2 snapshot
+	sh.reqsnapc <- struct{}{}
+	// TODO: generate v3 snapshot at here
+	snap := <-sh.snapc
+
+	d, err := snap.clone.SaveNoCopy()
+	if err != nil {
+		plog.Panicf("store save should never fail: %v", err)
+	}
+	return raftpb.Snapshot{
+		Metadata: snap.metadata,
+		Data:     d,
+	}
+}
+
+// SnapshotDataSize returns the size of latest snapshot data.
+func (sh *snapshotStore) SnapshotDataSize() uint64 { return 0 }
```

Comment:
this should not be a method on the store. I would expect to see snapshot.Size().